### ISSUE TITLE
Refreshable Materialized Views

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#33919]
**Use Case**

Automate periodic imports of data into ClickHouse. Automate batch jobs for data transformation.

**Requirements**

It should run a query with some schedule and atomically replace a table with the result.

Schedule can be aligned to calendar like "3 AM every Thursday" or not aligned like "30 seconds after previous import has been finished".

It can contain some concurrency limits like "don't do more than 5 jobs of this type concurrently". Limits should work within named "concurrency group". The job also can have numeric priority to control what to run first when resources are available.

If it was not possible to run the job due to concurrency limits or if server was not up, the setting should control if it needs to run as soon as possible or skip to the next update time if more than some ratio of time before next update already passed.

The settings for randomization to avoid DoS of the sources like "spread the next update time around three minutes".

Error recovery: do up to N retries with some initial timeout and exponential backoff saturated up to another timeout.

Chaining and dependencies between the jobs: run a job only after other specified jobs are finished.

System table for introspection of these jobs and system commands to temporary stop scheduling, or stop or pause already running job, or run the job out of order.

It should be possible to run clickhouse-server with all the jobs disabled before manual command.

**Out of scope**

Incremental data import into partitions by time. Batch jobs with multiple queries.
We can think how to extend this mechanism to something like "every day do import and replace partitions for three previous days".

Coordination of jobs in a cluster. (Should be easily extended with schedule in Keeper).

Backing up old tables after refresh (e.g. store N backups in renamed tables), can be easily implemented.

**Proposed solution**

```
CREATE MATERIALIZED VIEW name

REFRESH 

    EVERY 1 WEEK /* on Mondays */
    | EVERY 1 WEEK OFFSET 2 DAY 3 HOUR /* Wednedsday 3 AM */
    | EVERY 30 SECOND /* on 00 and 30 second every minute */
    | AFTER 30 SECOND /* unaligned, 30 seconds after previous update has been finished */
    
    RANDOMIZE FOR 3 MINUTE /* spread uniformly random plus minus 1.5 minutes */
    
DEPENDS ON other_view1, other_view2...  /* run after subsequent finish of these views, only for AFTER */
    
SETTINGS /* fine settings for a refresh process */
    max_concurrency = 3,
    concurrency_group = 'import_from_mysql',
    priority = 1,
    max_retries = 5,
    backoff_min_timeout = 10,
    backoff_max_timeout = 600,
    backoff_exponent = 2,
    skip_update_after_seconds = 1000,
    skip_update_after_ratio = 0.5

ENGINE = /* how to store data */

AS SELECT /* what query to run, possibly including the SETTINGS for the query */
```

```
SELECT * FROM system.view_refreshes;
SYSTEM REFRESH VIEW name;    /* ask for refresh out of schedule */
SYSTEM STOP|START VIEW name; /* do not run this view on schedule */
SYSTEM CANCEL|PAUSE|START VIEW name; /* for currently running job */
```

ALTER, DROP, RENAME, DETACH, ATTACH work trivially for these views.
ALTER is similar to CREATE and simply changes everything (the underlying table itself will be replaced on next update).

The underlying table has the same name as the view (the view and its data are represented with the same name in the database).

**Additional context**

With the help of Atomic databases, table replace is lock-free.

Refreshable materialized views will simplify managing some dictionaries. For example, we can create a view to import data from MySQL to EmbeddedRocksDB table and then create a `direct` dictionary on top of this table.

Note: this task is not hard to implement, but it contains many minor details and gotchas.

**Comment by UnamedRus** [comment]
Big problem for current POPULATE implementation and future refreshable MV is:

Source table is big as well as amount of distinct keys in MV, so it doesn't fit in memory in case of simple GROUP BY.
So we need to think about some solution, which will allow to create MV on top of really big tables and not kill (or overload) server at the same time.

**Comment by alexsubota** [comment]
Currently materialized views used for "live" update of data, can we change the name of this function? like "refreshable view", "scheduler view" or smth like that

**Comment by aramperes** [comment]
What would be the differences between this and [Experimental Live Views](https://clickhouse.com/docs/en/sql-reference/statements/create/view#live-view-experimental)?

It sounds like a refreshable materialized view with more options like `ENGINE` would remove the need for the `LIVE VIEW`?

**Comment by alexey-milovidov** [comment]
They are different:
- refreshable materialized views implement periodic updates, they can be used to pull data from external sources periodically, and process it, much like ELT;
- live views are subscribing to realtime updates of the tables to react on inserts instantly, but only work for local tables.

**Comment by aramperes** [comment]
@alexey-milovidov 

> * live views are subscribing to realtime updates of the tables to react on inserts instantly, but only work for local tables.

... not to tell you your business, but I think it's actually the opposite. I have a reproducable example in this repository: https://github.com/aramperes/grafana-clickhouse-jdbc, specifically https://github.com/aramperes/grafana-clickhouse-jdbc/blob/master/infra/clickhouse/clickhouse-init.sql:

```
629974a20971 :) CREATE TABLE comments_pg
                (
                    `id` Int32,
                    `by` String,
                    `parent` Int32,
                    `text` Nullable(String),
                    `time` DateTime64(6),
                    `story` Int32
                )
                ENGINE = JDBC('jdbc:postgresql://postgres:5432/hn?user=hn&password=password', 'public', 'comments');

CREATE TABLE comments_pg
(
    `id` Int32,
    `by` String,
    `parent` Int32,
    `text` Nullable(String),
    `time` DateTime64(6),
    `story` Int32
)
ENGINE = JDBC('jdbc:postgresql://postgres:5432/hn?user=hn&password=password', 'public', 'comments')

Query id: 2210cc3e-5d0c-4e84-b00f-2a3849b18c07

Ok.

0 rows in set. Elapsed: 0.018 sec.

629974a20971 :) CREATE LIVE VIEW comments
                WITH REFRESH 30
                AS SELECT
                  *
                FROM comments_pg;

CREATE LIVE VIEW comments WITH PERIODIC REFRESH 30 AS
SELECT *
FROM comments_pg

Query id: 47834bb0-d79b-4fa1-9486-807b10fb4351

Ok.

0 rows in set. Elapsed: 0.010 sec.

629974a20971 :) select count(*) from comments_pg;

SELECT count(*)
FROM comments_pg

Query id: f2ccf3d4-4f95-47f6-8437-ab5292e15dfd

┌─count()─┐
│      30 │
└─────────┘

1 row in set. Elapsed: 2.213 sec.

629974a20971 :) select count(*) from comments;

SELECT count(*)
FROM comments

Query id: f4ecb36b-cc8d-4317-aa0a-c9ad7163db54

┌─count()─┐
│      30 │
└─────────┘

1 row in set. Elapsed: 0.357 sec.
```

then, waiting 30 seconds later for a refresh,

```
629974a20971 :) select count(*) from comments;

SELECT count(*)
FROM comments

Query id: 361432e0-0953-4726-b4f7-eb64a0af5e51

┌─count()─┐
│      49 │
└─────────┘

1 row in set. Elapsed: 0.486 sec.
```

In fact, I originally tried `MATERIALIZED VIEW` with `comments_pg`, but as the docs say:

> Materialized views in ClickHouse are implemented more like insert triggers. If there’s some aggregation in the view query, it’s applied only to the batch of freshly inserted data

... when creating a `MATERIALIZED VIEW` on a JDBC table, there is no way for it to be updated, as the inserts happen in the external database. So unless I use `POPULATER`, a `MATIERALIZED VIEW` on a JDBC table never has any rows.

**Comment by alexey-milovidov** [comment]
Thank you! I did not use the periodic refresh feature before.

**Comment by R-omk** [comment]
https://clickhouse.com/docs/en/sql-reference/dictionaries#dictionary-updates

I just want to pay attention.

**Comment by aramperes** [comment]
@R-omk That's also a good reminder. Here's the comparison between the 3 solutions as I see it:

<table>
<tr>
	<td>
	<td><code>DICTIONARY</code> with <code>LIFETIME</code>
	<td><code>LIVE VIEW</code> with <code>REFRESH</code>
	<td><code>MATERIALIZED VIEW</code> with <code>REFRESH</code>
<tr>
	<td>Storage Options
	<td>Memory (<a href="https://clickhouse.com/docs/en/sql-reference/dictionaries#ways-to-store-dictionaries-in-memory">configurable</a>, also <code>ssd_cache</code>)
	<td>Memory (<a href="https://github.com/ClickHouse/ClickHouse/blob/f1df91cd0998ee7ba368972855bf16c84d920a9f/src/Storages/LiveView/StorageLiveView.cpp">hash?</a>)
	<td>Configurable Engine (disk/memory)
<tr>
	<td>Insert <code>TO</code> Tables
	<td>
	<td>
	<td>✔️
<tr>
	<td>Invalidation Query
	<td>✔️
	<td>
	<td>
<tr>
	<td>Local Table Source
	<td>
	<td>✔️
	<td>✔️
<tr>
	<td>DBMS Table Source
	<td><a href="https://clickhouse.com/docs/en/sql-reference/dictionaries#dbms">Limited</a>
	<td>✔️
	<td>✔️
<tr>
	<td>Concurrency Groups
	<td>
	<td>
	<td>✔️
</table>

**Comment by aramperes** [comment]
Until #46317 is merged, would it technically be feasible to make a `LIVE VIEW WITH REFRESH` trigger inserts into a `MATERIALIZED TABLE` (the same way `LIVE VIEW WITH REFRESH` dispatches updates to `WATCH`?)

```sql
CREATE TABLE source_table (`id` Int32, `text` String) ENGINE = Memory;

CREATE LIVE VIEW live_view WITH REFRESH 30 AS SELECT id, text from source_table;

CREATE MATERIALIZED VIEW materialized (`id` Int32, `text` String) ENGINE = Memory
AS SELECT id, text FROM live_view;
```

```sql
INSERT INTO source_table (id, text) VALUES (1, 'test');

Received exception from server (version 23.2.4):
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Block structure mismatch in function connect between PushingToLiveViewSink and CopyingDataToViewsTransform stream: different number of columns:
id Int32 Int32(size = 0), text String String(size = 0)
id Int32 Int32(size = 0), text String String(size = 0), _version UInt64 UInt64(size = 0). (LOGICAL_ERROR)
```

The error seems like a bug &mdash; `_version` is a block column normally propagated to `WATCH` but the `CopyingDataToViewsTransform` doesn't seem to understand it.

**Comment by arvindpunk** [comment]
> when creating a `MATERIALIZED VIEW` on a JDBC table, there is no way for it to be updated, as the inserts happen in the external database. So unless I use `POPULATE`, a `MATIERALIZED VIEW` on a JDBC table never has any rows.

How about running an (external) cron that refreshes source table using clickhouse local
```sql
clickhouse local --query "
INSERT INTO FUNCTION
remote('localhost', '[db.]target_table')
SELECT * FROM jdbc('custom_source', 'SELECT * FROM [db.]source_table');"
```

That way you'd have a continuously up to date data in ClickHouse. Yes, a CH native solution (the WIP PR) would be amazing, but until then, this seems like a viable option as well.

**Comment by alexey-milovidov** [comment]
@arvindpunk Yes, it is a totally viable option!

**Comment by simonmahns** [comment]
wen

**Comment by spelwisselaar** [comment]
@alexey-milovidov I am using clickhouse 22.2.2.1. My requirement is to fetch records from multiple tables (around 17 to 18 tables with join conditions) in real time. I tried to create MV with the refresh keyword, but I am getting exceptions. Am I doing anything wrong?

For reference, I am attaching DDL of one test MV

CREATE MATERIALIZED VIEW refresh_view
REFRESH AFTER 30 SECOND
( id Int32, full_name String, dob_1 Date, dob_2 Date, value UInt64)
ENGINE = ReplacingMergeTree
ORDER BY id
SETTINGS index_granularity = 8192 populate AS
SELECT a.id, concat(a.Name,' ', b.Name) AS full_name, a.dob AS dob_1, b.dob AS dob_2, b.value FROM PROJ_B01.replacing_test AS a LEFT JOIN PROJ_B01.replacing_test_2 AS b ON b.id = a.id;

**Comment by arctica** [comment]
@aramperes I found that you can make your table -> live view -> materialized view method work by either using `SELECT * EXCEPT(_version) FROM live_view` or for example `SELECT id, text FROM (SELECT id, text FROM live_view)`. Well, to be precise it got rid of the error and made the data show up in the materialized view but it seems like the live view didn't adhere to the refresh timer and just passed it on immediately as if the materialized view was attached to the source_table directly.

**Comment by eustrainLee** [comment]
I noticed that the implementation uses the std::chrono::system_clock as a basis for determining current time. How can I avoid the impact of updating logic due to local time updates?

**Comment by alexey-milovidov** [comment]
@al13n321, let's change to a monotonic clock.

**Comment by al13n321** [comment]
`REFRESH AFTER` can use monotonic clock, but `REFRESH EVERY 1 DAY` means "refresh at midnight", so it needs to know when midnight is...

**Comment by alexey-milovidov** [comment]
@eustrainLee, I think it's not a big deal - we can keep it as is.

**Comment by al13n321** [comment]
> I noticed that the implementation uses the std::chrono::system_clock as a basis for determining current time. How can I avoid the impact of updating logic due to local time updates?

For small adjustments nothing bad should happen.

The only bad case is if the clock was decreased by a lot. E.g. suppose we did a refresh at 2050-01-01, then the clock was moved back to 2025-01-01; the next refresh will be scheduled for 25+ years later, i.e. approximately never. But a server restart would unbreak it.

(Can small adjustment cause a refresh to happen twice? E.g. the clock reached 03:00:00 causing a refresh, then the clock was moved back to 02:59:59 and reached 03:00:00 again. No, that won't trigger another refresh, the next refresh is scheduled for 03:00:00 + period.)
(Can daylight savings time cause problems? No, everything is in UTC.)
